torch/functional.py

```diff
@@ -1235,16 +1235,29 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
             None   Frobenius norm                2-norm
             'fro'  Frobenius norm                --
             'nuc'  nuclear norm                  --
-            Other  as vec norm when dim is None  sum(abs(x)**ord)**(1./ord)
+            Other  --                            sum(abs(x)**ord)**(1./ord)
             =====  ============================  ==========================
 
-        dim (int, 2-tuple of ints, 2-list of ints, optional): If it is an int,
-            vector norm will be calculated, if it is 2-tuple of ints, matrix norm
-            will be calculated. If the value is None, matrix norm will be calculated
-            when the input tensor only has two dimensions, vector norm will be
-            calculated when the input tensor only has one dimension. If the input
-            tensor has more than two dimensions, the vector norm will be applied to
-            last dimension.
+            The vector norm can be calculated across any number of dimensions.
+            The corresponding dimensions of :attr:`input` are flattened into
+            one dimension, and the norm is calculated on the flattened
+            dimension. The vector norm will be used if :attr:`p` is a number,
+            or if :attr:`p` is ``None`` and :attr:`input` has either one or
+            more than three dimensions.
+
+            Frobenius norm and nuclear norm can only be calculated across two
+            dimensions.
+
+            When ``p='fro'`` is set, and either :attr:`input` or :attr:`dim`
+            only specifies one dimension, the Frobenius norm will not actually
+            be calculated. Instead, the vector 2-norm will be calculated.
+
+        dim (int, tuple of ints, list of ints, optional):
+            Specifies which dimension or dimensions of :attr:`input` to
+            calculate the norm across. If :attr:`dim` is ``None``, the norm will
+            be calculated across all dimensions of :attr:`input`. If the norm
+            type indicated by :attr:`p` does not support the specified number of
```

Comment:
I guess I'm confused by this:

```
t = torch.randn((3, 3, 3))
torch.norm(t)
```

This works. But the tensor has three dimensions. dim is None, so the norm is computed across all the dimensions. And p is 'fro' so the Frobenius norm should be computed. But the Frobenius norm is a matrix norm, so I would think it would fail in this case (or run batched over each matrix)?

What's going on here?

Reply:
Yeah you're right. I was wrong--`p='fro'` (the default setting) will accept any input dimensionality.

Yes, Frobenius norm is supposed to be just a matrix norm, so it should only accept 2-D inputs (https://en.wikipedia.org/wiki/Matrix_norm#Frobenius_norm). The API semantics for `p='fro'` seem to be misaligned with the mathematical definition for Frobenius norm. As it stands, I believe `torch.norm`'s behavior for `p='fro'` and `p=2` are equivalent:

```
>>> a = torch.randn(100)
>>> torch.norm(a, p='fro') == torch.norm(a, p=2)
tensor(True)
>>> a = torch.randn(100, 20)
>>> torch.norm(a, p='fro') == torch.norm(a, p=2)
tensor(True)
>>> a = torch.randn(100, 20, 300)
>>> torch.norm(a, p='fro') == torch.norm(a, p=2)
tensor(True)
```

I'll update the documentation to point this out.

Note that `numpy.linalg.norm` and `torch.linalg.norm` have mathematically correct behavior (at least as far as I understand) because they only accept 2-D reductions:

```
>>> torch.norm(torch.randn(3), p='fro')
tensor(2.0952)
>>> torch.norm(torch.randn(3, 4), p='fro')
tensor(3.8539)
>>> torch.norm(torch.randn(3, 4, 5), p='fro')
tensor(8.6134)
>>> torch.linalg.norm(torch.randn(3), ord='fro')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: order "fro" can only be used if either len(dim) == 2 or (self.dim() == 2 and dim is None)
>>> torch.linalg.norm(torch.randn(3, 4), ord='fro')
tensor(3.6517)
>>> torch.linalg.norm(torch.randn(3, 4, 5), ord='fro')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: order "fro" can only be used if either len(dim) == 2 or (self.dim() == 2 and dim is None)
```

Reply:
Ah ok, this is why I thought `p='fro'` only supports one or two dimensions:

```
>>> a = torch.randn(2, 3, 4)
>>> a.norm(p='fro', dim=0)
tensor([[1.3950, 1.0560, 0.9329, 2.9083],
        [1.4607, 1.3968, 0.7379, 0.2507],
        [1.2769, 1.0869, 0.7636, 0.9489]])
>>> a.norm(p='fro', dim=(0, 1))
tensor([2.3896, 2.0609, 1.4134, 3.0695])
>>> a.norm(p='fro', dim=(0, 1, 2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/endoplasm/miniconda3/lib/python3.7/site-packages/torch/tensor.py", line 339, in norm
    return torch.norm(self, p, dim, keepdim, dtype=dtype)
  File "/home/endoplasm/miniconda3/lib/python3.7/site-packages/torch/functional.py", line 755, in norm
    return torch._C._VariableFunctions.frobenius_norm(input, dim, keepdim=keepdim)
RuntimeError: Expected at most 2 dimensions, but got 3 dimensions instead.
```

It only supports more than 2 dimensions if `dim=None`. I'll make note of that.

Reply:
Alright, I think my update describes the behavior more precisely, without contradiction. Let me know what you think @mruberry .